File: crates/dewasm-backend-go/units/rt/i32_trunc_sat_u.go
func (rt) i32_trunc_sat_u(x float64) uint32 {
    if math.IsNaN(x) {
        return 0
    }
    t := math.Trunc(x)
    if t < 0 {
        return 0
    }
    if t > 4294967295 {
        return 4294967295
    }
    return uint32(t)
}

File: crates/dewasm-backend-java/units/rt/i32_trunc_sat_u.java
// i32.trunc_sat_f32_u / i32.trunc_sat_f64_u: saturating unsigned truncation.
// Java's `(int)(long)` would wrap past u32, so saturation is explicit (NaN->0, clamp to [0, 2^32)); the signed saturating forms use a plain `(int)` cast, which already matches wasm.
static int i32_trunc_sat_u(double x) {
    if (Double.isNaN(x)) {
        return 0;
    }
    double t = x < 0 ? Math.ceil(x) : Math.floor(x);
    if (t <= 0) {
        return 0;
    }
    if (t >= 4294967295.0) {
        return -1;
    }
    return (int) (long) t;
}
